csfieldguide/static/interactives/city-trip/js/city-trip.js

```diff
@@ -0,0 +1,642 @@
+    return {distance: totalDistance, isBestRoute: true};
+  } else {
+    return {distance: totalDistance, isBestRoute: false};
+  }
+}
+
+
+/** Finds the edge differences between two paths 
+ *  and draws the new path by calling changePaths */
+function findEdgeDifferences(cyGraph, newEdgeConfig, numCities, startNode) {
+  // Remove edge that closes the loop
+  var lastNodeID = numCities.toString();
+  cyGraph.remove(cyGraph.$('#' + lastNodeID + '-' + startNode));
+  // Number of nodes (cities) remains the same between paths;
+  var oldEdgeConfig = new Set();
+  for (var i = 0; i < numCities - 1; i++) {
+    var edgeID = cyGraph.elements('edge')[i].data('id');
+    oldEdgeConfig.add(edgeID);
+  }
+
+  var edgesToRemove = setDifference(oldEdgeConfig, newEdgeConfig);
+  var edgesToAdd = setDifference(newEdgeConfig, oldEdgeConfig);
+  var edgesToKeep = setDifference(oldEdgeConfig, edgesToRemove);
+  // Draw the next path;
+  changePaths(cyGraph, edgesToKeep, edgesToAdd);
+}
+
+
+/** Draws the path */
+function changePaths(cy, edgesToKeep, edgesToAdd) {
+  var edgesToKeepStr = '';
+  var edgesToKeepArr = Array.from(edgesToKeep);
+  for (var i = 0; i < edgesToKeepArr.length - 1; i++) {
+    edgesToKeepStr += '#' + edgesToKeepArr[i] + ', ';
+  }
+  if (edgesToKeepArr.length > 0) {
+    edgesToKeepStr += '#' + edgesToKeepArr[i];
+  }
+  cy.remove(cy.edges().difference(edgesToKeepStr));
+
+  for (let edgeId of edgesToAdd) {
+    edgeNodes = edgeId.split('-');
+    var sourceNode = edgeNodes[0];
+    var targetNode = edgeNodes[1];
+    addEdge(cy, sourceNode, targetNode);
+  }
+}
+
+
+/** Calulate a\b 
+ *  E.g if a = {1,2,3,4} and b = {5,4,3,2} this function would return {1} */
+function setDifference(a, b) {
+  var aMinusB = new Set([...a].filter(x => !b.has(x)));
+  return aMinusB;
+}
+
+
+/** E.g if the startingCity is 1 and cities is [2,3] return [1,2,3,1] */
+function addStartingCityToPath(startingCity, cities) {
+  // Add origin city to start of route
+  temp = [startingCity].concat(cities);
+  // Add origin city to end of route and return
+  return temp.concat(startingCity);
+}
+
+
+/** Swap elements at the ith and jth indexes with each other and return. */
+function swap(A, i, j) {
+  [A[i], A[j]] = [A[j], A[i]];
+  return A;
+}
+
+
+  /** Get all permutations of the intermediate cities one by one.
+   *  Show the path on the graph as each one if found.
+   *  Based upon the non-recursive version of Heap's algorithm. 
+   *  Non-recursive version used so we can draw the new path as soon as it is generated.
+   */
+async function permutations(cy, cy2, intermediateCities, bestRouteDistance, startingCity, k) {
+  // c is an encoding of he stack state
+  c = new Array(k).fill(0);
+  pathWithStartingCity = addStartingCityToPath(startingCity, intermediateCities);
+
+  // Draw graph here
+  pathData = testNewPath(cy, cy2, pathWithStartingCity, bestRouteDistance, startingCity.toString());
+  await sleep(100);
```

Comment:
This is an interesting problem I learned about in an ENCE class. This is a delay of 0.1s *plus* the delay to run the code itself. On my low-end (but probably reasonably representative) PC it's only 40s before the interactive is an entire second behind schedule. There are ways around this but I don't think it matters enough in this situation to warrant changing anything